### PR TITLE
Derive macro for tagged enums (GraphQLUnion)

### PR DESCRIPTION
**Pull request by jmpunkt**
Implements derive macros for tagged enums which are translated into GraphQLUnion types. The following features are provided

- The macro implements GraphQLAsyncType. 
- If all variants of the tagged enum have different types, then the derive macro implements `std::convert::From<T>`  converter for these types.

Fixes: https://github.com/graphql-rust/juniper/issues/586

### CODE REVIEW

File: juniper/CHANGELOG.md [hunk 1]
```diff
@@ -20,6 +20,11 @@ See [#419](https://github.com/graphql-rust/juniper/pull/419).
 
 See [#569](https://github.com/graphql-rust/juniper/pull/569).
 
+- GraphQLUnion derive support ("#[derive(GraphqQLUnion)]")
+  - implements GraphQLAsyncType
+
+See [#](https://github.com/graphql-rust/juniper/pull/).
```
Comment (LegNeato) [inline juniper/CHANGELOG.md:26]: Nit: We can fill this out now 😃

File: juniper_codegen/src/derive_union.rs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    // Parse attributes.
+    let attrs = match util::ObjectAttributes::from_attrs(&ast.attrs) {
+        Ok(a) => a,
+        Err(e) => {
+            panic!("Invalid #[graphql(...)] attribute for enum: {}", e);
+        }
+    };
+
+    if !attrs.interfaces.is_empty() {
+        panic!("#[derive(GraphlQLUnion)] does not support interfaces");
```
Comment (LegNeato) [inline juniper_codegen/src/derive_union.rs:24]:
```suggestion
        panic!("#[derive(GraphQLUnion)] does not support interfaces");
```

File: juniper_codegen/src/derive_union.rs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            let _type = match field.fields {
+                Fields::Unnamed(inner) => {
+                    let mut iter = inner.unnamed.iter();
+                    let first = match iter.next() {
+                        Some(val) => val,
+                        None => unreachable!(),
+                    };
+
+                    if iter.next().is_some() {
+                        panic!("#[derive(GraphlQLUnion)] all members must be unnamed with a single element e.g. Some(T)");
```
Comment (LegNeato) [inline juniper_codegen/src/derive_union.rs:62]:
```suggestion
                        panic!("#[derive(GraphQLUnion)] all members must be unnamed with a single element e.g. Some(T)");
```

File: juniper_codegen/src/derive_union.rs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+                    }
+
+                    first.ty.clone()
+                }
+                _ => panic!("#[derive(GraphlQLObject)] all fields of the enum must be unnamed"),
```
Comment (LegNeato) [inline juniper_codegen/src/derive_union.rs:67]:
```suggestion
                _ => panic!("#[derive(GraphQLUnion)] all fields of the enum must be unnamed"),
```

**Comment by jmpunkt** [comment]
Thanks for the review. Everything you mentioned is fixed. During the testing i encountered some issues. Therefore, skip and documentation annotations are disabled for enumeration fields. Here are some thoughts about these decisions and the implementation. If something is unclear or wrong, let me know.

- Documentation of variations is not possible in GraphQL? (not 100% sure)
- Skipping fields works in general but throws panics during runtime. For example, if a query returns a skipped variant of the enum. The derive macro should provide additional safety (compared to the other union implementation). If the user requires this behavior he can fallback to the other union implementation.
- Currently the std::convert::Into<T> ensures that each variant has a unique type. Due to the GraphQL specification this must be done? (not 100% sure)

File: docs/book/content/types/unions.md [hunk 5]
```diff
@@ -187,3 +188,34 @@ impl Character {
...
 # fn main() {}
 ```
+
+## Enums (Derive)
+
+This example is similar to `Enums (Impl)`. To successfully use the
+derive macro, ensure that each variant of the enum has a different
+type. Since each variant is different, the device macro provides
+`std::convert::Into<T>` converter for each _used_ variant. Fields
+which are skipped by using `#[graphql(skip)]` do not have a
```
Comment (LegNeato) [inline docs/book/content/types/unions.md:198]: This is no longer true, correct? It appears below you explicitly forbid skip for this (which feels right).

**Comment by jmpunkt** [comment]
This PR should be ready to merge. Applied your feedback and added some minor tweaks.

**Comment by LegNeato** [comment]
Awesome, thank you again!

**Pull request merged**